File: kometa.py
#!/usr/bin/env python3
# ─────────────────────────────────────────────
# kometa.py
# Homoglyph steganography with password-derived keys.
# Zero dependencies — standard library only.
#
# Usage:
#   python3 kometa.py encode <cover> <message> <output>
#   python3 kometa.py decode <input> <output>
# ─────────────────────────────────────────────

import sys, os, hashlib, hmac, struct, getpass

# ── CONSTANTS & CONFIG ────────────────────────

SEED_BYTES       = 8
BLOCK_SIZE       = 64
SALT_COVER_BYTES = 4096

# Adjust memalloc to 1<<15 to improve performance.
# Tradeoff is brute-force cost: compromise expensive but possible.
SCRYPT = dict(n=1<<17, r=8, p=1, dklen=None)  # dklen set per-call

# ── DISTRIBUTION CONFIG ───────────────────────

DEAD_ZONE_START  = 0.10   # fraction of carriers to exclude at head
DEAD_ZONE_END    = 0.10   # fraction of carriers to exclude at tail
DENSITY_BUCKETS  = 20     # granularity of carrier density profile

# Homoglyph pairs — positional: lat[n] is the visual twin of cyr[n].
# Ordered by frequency/importance; mnemonic: KOMETA = same in Latin & Cyrillic.
DICT = dict(
    lat = "acepoijxyKOMETAPHIJX",
    cyr = "асероіјхуКОМЕТАРНІЈХ",
)

# ── KEY DERIVATION ────────────────────────────

# Build char → index lookup for normalisation (covers both scripts)
_NORM = {ch: i for i, ch in enumerate(DICT["cyr"])}
_NORM.update({ch: i for i, ch in enumerate(DICT["lat"])})

def _normalise(text: str) -> str:
    """Replace all homoglyph carriers with their Latin equivalents."""
    return "".join(DICT["lat"][_NORM[ch]] if ch in _NORM else ch for ch in text)

def _derive_bootstrap(password: str, cover: str) -> bytes:
    """One scrypt call — returns 64 bytes of base key material."""
    salt   = hashlib.sha256(_normalise(cover)[:SALT_COVER_BYTES].encode()).digest()
    maxmem = 128 * SCRYPT["n"] * SCRYPT["r"] * 2
    return hashlib.scrypt(password.encode(), salt=salt, dklen=64,
                          n=SCRYPT["n"], r=SCRYPT["r"], p=SCRYPT["p"], maxmem=maxmem)

def _bootstrap_seed(base_key: bytes) -> bytes:
    """Seed for nonce pool — first 8 bytes of base key, no nonce involved."""
    return base_key[:SEED_BYTES]

def derive_keys(base_key: bytes, message_len: int, nonce: bytes) -> tuple[bytes, bytes]:
    """Expand base key with nonce via HKDF-SHA256. Returns (seed[8], keystream[n])."""
    keylen = ((SEED_BYTES + message_len + BLOCK_SIZE - 1) // BLOCK_SIZE) * BLOCK_SIZE
    # HKDF-expand: cheap HMAC-based expansion, no second scrypt
    prk  = hmac.digest(nonce, base_key, "sha256")  # extract step (nonce as salt)
    out  = b""
    prev = b""
    for i in range(1, (keylen // 32) + 2):
        prev = hmac.digest(prk, prev + i.to_bytes(1, "big"), "sha256")
        out += prev
    key = out[:keylen]
    return key[:SEED_BYTES], key[SEED_BYTES:]

# ── PRIMITIVES ────────────────────────────────

def _xor(data: bytes, keystream: bytes) -> bytes:
    return bytes(a ^ b for a, b in zip(data, keystream))

def _to_bits(data: bytes) -> list[int]:
    if len(data) > 65535: raise ValueError("Message too long (max 65,535 bytes)")
    framed = bytes([len(data) >> 8, len(data) & 0xff]) + data
    return [(b >> i) & 1 for b in framed for i in range(7, -1, -1)]

def _from_bits(bits: list[int]) -> bytes:
    def read_byte(offset):
        v = 0
        for b in range(8): v = (v << 1) | (bits[offset + b] if offset + b < len(bits) else 0)
        return v
    length = (read_byte(0) << 8) | read_byte(8)
    return bytes(read_byte(16 + i * 8) for i in range(length))

# xoshiro128**
def _make_rng(seed: bytes):
    s0, s1 = struct.unpack_from("<II", seed, 0)
    s0 |= 1; s1 |= 1
    s2, s3 = 0x9e3779b9, 0x6c62272e
    def rotl(x, k): return ((x << k) | (x >> (32 - k))) & 0xFFFFFFFF
    def rng():
        nonlocal s0, s1, s2, s3
        r = (rotl(s1 * 5 & 0xFFFFFFFF, 7) * 9) & 0xFFFFFFFF
        t = (s1 << 9) & 0xFFFFFFFF
        s2 ^= s0; s3 ^= s1; s1 ^= s2; s0 ^= s3; s2 ^= t
        s3 = rotl(s3, 11)
        return r
    return rng

def _shuffle(arr: list, rng) -> list:
    a = arr[:]
    for i in range(len(a) - 1, 0, -1):
        j = rng() % (i + 1)
        a[i], a[j] = a[j], a[i]
    return a

# ── STEGANOGRAPHY ─────────────────────────────

# Build char → index lookup for all carriers (both scripts, flat)
_LOOKUP = {}
for _i, (_l, _c) in enumerate(zip(DICT["lat"], DICT["cyr"])):
    _LOOKUP[_l] = ("lat", _i)
    _LOOKUP[_c] = ("cyr", _i)

def _nonce_positions(positions: list[int], seed: bytes) -> list[int]:
    """Select 64 fixed carrier positions for the nonce (drawn before header/body)."""
    n  = len(positions)
    lo = int(n * DEAD_ZONE_START)
    hi = n - int(n * DEAD_ZONE_END)
    eligible = positions[lo:hi]
    if len(eligible) < 64:
        raise ValueError(f"Cover too small: need at least 64 carriers for nonce, got {len(eligible)}")
    return _shuffle(eligible, _make_rng(seed))[:64]

def _header_positions(positions: list[int], seed: bytes, bootstrap_seed: bytes) -> list[int]:
    """
    Select 16 fixed carrier positions for the length header.
    Drawn from the dead-zone-excluded pool via a seeded shuffle —
    independent of message length, so decode can always recover them
    without knowing n_bits first. Header positions are excluded from
    body selection.
    """
    n  = len(positions)
    lo = int(n * DEAD_ZONE_START)
    hi = n - int(n * DEAD_ZONE_END)
    nonce_set = set(_nonce_positions(positions, bootstrap_seed))
    eligible = [p for p in positions[lo:hi] if p not in nonce_set]
    if len(eligible) < 16:
        raise ValueError(f"Cover too small: need at least 16 eligible carriers, got {len(eligible)}")
    return _shuffle(eligible, _make_rng(seed))[:16]

def _body_positions(positions: list[int], header_pos: list[int], n_bits: int, seed: bytes,
                    bootstrap_seed: bytes, n_buckets: int = DENSITY_BUCKETS) -> list[int]:
    """
    Select n_bits carrier positions for the message body using
    density-matched bucket allocation. Header and nonce positions are excluded.
    """
    n  = len(positions)
    lo = int(n * DEAD_ZONE_START)
    hi = n - int(n * DEAD_ZONE_END)
    nonce_set  = set(_nonce_positions(positions, bootstrap_seed))
    header_set = set(header_pos)
    eligible   = [p for p in positions[lo:hi] if p not in nonce_set and p not in header_set]

    if len(eligible) < n_bits:
        raise ValueError(
            f"Cover too small after dead-zone trim: "
            f"need {n_bits} body carriers, have {len(eligible)} eligible "
            f"(total {n}, margins {DEAD_ZONE_START:.0%}/{DEAD_ZONE_END:.0%})"
        )

    # ── density-matched bucket allocation ────
    actual_buckets = min(n_buckets, len(eligible))
    bucket_size    = len(eligible) / actual_buckets
    buckets        = [
        eligible[int(i * bucket_size) : int((i + 1) * bucket_size)]
        for i in range(actual_buckets)
    ]

    raw   = [len(b) / len(eligible) * n_bits for b in buckets]
    alloc = [int(r) for r in raw]
    remainder = n_bits - sum(alloc)
    fracs = sorted(range(actual_buckets), key=lambda i: -(raw[i] - alloc[i]))
    for i in fracs[:remainder]:
        alloc[i] += 1

    rng      = _make_rng(seed)
    selected = []
    for bucket, k in zip(buckets, alloc):
        chosen = _shuffle(bucket, rng)[:k]
        selected.extend(chosen)

    return selected

def _embed(cover: str, bits: list[int], bootstrap_seed: bytes, seed: bytes, nonce: bytes) -> str:
    # bits layout: first 16 = length header, rest = body
    # lat = bit 0 (inactive), cyr = bit 1 (active)
    header_bits = bits[:16]
    body_bits   = bits[16:]

    positions = [i for i, ch in enumerate(cover) if ch in _LOOKUP]
    nonce_pos = _nonce_positions(positions, bootstrap_seed)
    hdr_pos   = _header_positions(positions, seed, bootstrap_seed)
    body_pos  = _body_positions(positions, hdr_pos, len(body_bits), seed, bootstrap_seed)

    nonce_bits = [(nonce[i // 8] >> (7 - i % 8)) & 1 for i in range(64)]

    active = {}
    for p, b in zip(nonce_pos, nonce_bits): active[p] = b
    for p, b in zip(hdr_pos, header_bits):  active[p] = b
    for p, b in zip(body_pos, body_bits):   active[p] = b

    out = []
    for i, ch in enumerate(cover):
        if ch not in _LOOKUP:
            out.append(ch)
        else:
            idx = _LOOKUP[ch][1]
            bit = active.get(i, 0)                  # inactive positions → bit 0
            out.append(DICT["cyr"][idx] if bit else DICT["lat"][idx])
    return "".join(out)

def _extract_nonce(encoded: str, bootstrap_seed: bytes) -> bytes:
    """Read the 64-bit nonce from its fixed carrier pool using the bootstrap seed."""
    positions = [i for i, ch in enumerate(encoded) if ch in _LOOKUP]
    nonce_pos = _nonce_positions(positions, bootstrap_seed)
    observed  = {i for i, ch in enumerate(encoded) if ch in _LOOKUP and _LOOKUP[ch][0] == "cyr"}
    bits = [(1 if p in observed else 0) for p in nonce_pos]
    return bytes(sum(bits[i*8+b] << (7-b) for b in range(8)) for i in range(8))

def _extract(encoded: str, seed: bytes, bootstrap_seed: bytes) -> list[int]:
    positions, observed = [], {}
    for i, ch in enumerate(encoded):
        if ch not in _LOOKUP: continue
        positions.append(i)
        if _LOOKUP[ch][0] == "cyr": observed[i] = 1

    def read_bit(p): return observed.get(p, 0)
    def read_byte(ps, offset):
        return sum(read_bit(ps[offset + b]) << (7 - b) for b in range(8))

    hdr_pos       = _header_positions(positions, seed, bootstrap_seed)
    payload_bytes = (read_byte(hdr_pos, 0) << 8) | read_byte(hdr_pos, 8)
    n_body_bits   = payload_bytes * 8

    n  = len(positions)
    lo = int(n * DEAD_ZONE_START)
    hi = n - int(n * DEAD_ZONE_END)
    nonce_set     = set(_nonce_positions(positions, bootstrap_seed))
    hdr_set       = set(hdr_pos)
    max_body_bits = len([p for p in positions[lo:hi] if p not in nonce_set and p not in hdr_set])
    n_body_bits   = min(n_body_bits, max(0, max_body_bits))

    body_pos    = _body_positions(positions, hdr_pos, n_body_bits, seed, bootstrap_seed)
    header_bits = [read_bit(p) for p in hdr_pos]
    body_bits   = [read_bit(p) for p in body_pos]
    return header_bits + body_bits

# ── ENCODE / DECODE ───────────────────────────

def _read_or_literal(arg: str) -> bytes:
    if os.path.exists(arg):
        with open(arg, "rb") as f: return f.read()
    return arg.encode()

def encode(cover_file: str, message_arg: str, password: str, output_file: str):
    cover   = open(cover_file, encoding="utf-8").read().rstrip()
    message = _read_or_literal(message_arg)
    nonce   = os.urandom(8)
    sys.stderr.write("⏳ Deriving keys…\n")
    base_key         = _derive_bootstrap(password, cover)
    bs               = _bootstrap_seed(base_key)
    seed, keystream  = derive_keys(base_key, len(message), nonce)
    encrypted = _xor(message, keystream)
    output    = _embed(cover, _to_bits(encrypted), bs, seed, nonce)
    open(output_file, "w", encoding="utf-8").write(output)
    sys.stderr.write(f"✓ Encoded → {output_file}\n")

def decode_text(encoded: str, password: str) -> bytes:
    """Pure in-memory decode — no disk I/O. Returns decrypted payload as bytes."""
    encoded  = encoded.rstrip()
    base_key = _derive_bootstrap(password, encoded)
    bs       = _bootstrap_seed(base_key)
    nonce    = _extract_nonce(encoded, bs)
    seed, _  = derive_keys(base_key, BLOCK_SIZE, nonce)
    bits      = _extract(encoded, seed, bs)
    encrypted = _from_bits(bits)
    _, keystream = derive_keys(base_key, len(encrypted), nonce)
    return _xor(encrypted, keystream)

def decode(input_file: str, password: str, output_file: str):
    """CLI wrapper — reads file, decodes in memory, writes result."""
    sys.stderr.write("⏳ Deriving keys…\n")
    encoded   = open(input_file, encoding="utf-8").read()
    decrypted = decode_text(encoded, password)
    open(output_file, "wb").write(decrypted)
    sys.stderr.write(f"✓ Decoded → {output_file}\n")

# ── CLI ───────────────────────────────────────

def _get_password() -> str:
    """Read password from KOMETA_PASSWORD env var (testing) or a secure prompt (interactive)."""
    env = os.environ.get("KOMETA_PASSWORD")
    if env is not None:
        return env
    return getpass.getpass("Password: ")

if __name__ == "__main__":
    args = sys.argv[1:]
    if len(args) == 4 and args[0] == "encode":
        encode(args[1], args[2], _get_password(), args[3])
    elif len(args) == 3 and args[0] == "decode":
        decode(args[1], _get_password(), args[2])
    else:
        sys.stderr.write(
            "usage:\n"
            "  python3 kometa.py encode <cover> <message> <output>\n"
            "  python3 kometa.py decode <input>  <output>\n"
            "\n"
            "Password prompted securely: never passed in argv\n"
        )
        sys.exit(1)

File: kometa.ts
// -----------------------------------------------
// kometa.ts
// Homoglyph steganography with password-derived keys.
// Zero npm dependencies -- Web Crypto API + plain TS only.
//
// Port notes vs. the original kometa.py:
//   - scrypt -> PBKDF2-SHA256 via crypto.subtle. Browsers don't ship scrypt
//     natively, and hand-rolling it would mean rolling our own crypto for
//     the one piece where that's genuinely dangerous. PBKDF2 is CPU-hard,
//     not memory-hard like scrypt, so it's weaker against GPU/ASIC brute
//     force at equal wall-clock cost. We compensate with a high iteration
//     count (PBKDF2_ITERATIONS below). This means encoded output is NOT
//     cross-compatible with kometa.py -- the key material differs.
//   - xoshiro128**, bit-packing, and carrier placement are ported as plain
//     TS: this is bespoke logic nobody else has audited but us, so
//     readability wins over hiding it behind a library.
//   - Sensitive byte buffers (keys, keystreams, password bytes) are
//     zeroed with .fill(0) as soon as they're no longer needed.
// -----------------------------------------------

// ---- CONSTANTS & CONFIG -----------------------

const SEED_BYTES = 8;
const BLOCK_SIZE = 64;
const SALT_COVER_CHARS = 4096; // character count (post-normalise), not bytes

// PBKDF2-SHA256 substituted for scrypt (see header note). OWASP's current
// floor for PBKDF2-SHA256 is 600,000 iterations; bump this over time.
const PBKDF2_ITERATIONS = 600_000;

// ---- DISTRIBUTION CONFIG -----------------------

const DEAD_ZONE_START = 0.10; // fraction of carriers excluded at head
const DEAD_ZONE_END = 0.10;   // fraction of carriers excluded at tail
const DENSITY_BUCKETS = 20;   // granularity of carrier density profile

// Homoglyph pairs -- positional: LAT[n] is the visual twin of CYR[n].
// Mnemonic: KOMETA = same in Latin & Cyrillic.
const LAT = "acepoijxyKOMETAPHIJX";
const CYR = "\u0430\u0441\u0435\u0440\u043e\u0456\u0458\u0445\u0443\u041a\u041e\u041c\u0415\u0422\u0410\u0420\u041d\u0406\u0408\u0425";

// ---- LOOKUP TABLES (built once at module load) ----

type Script = "lat" | "cyr";

const NORM = new Map<string, number>();
for (let i = 0; i < CYR.length; i++) NORM.set(CYR[i], i);
for (let i = 0; i < LAT.length; i++) NORM.set(LAT[i], i);

const LOOKUP = new Map<string, { script: Script; idx: number }>();
for (let i = 0; i < LAT.length; i++) {
  LOOKUP.set(LAT[i], { script: "lat", idx: i });
  LOOKUP.set(CYR[i], { script: "cyr", idx: i });
}

function normalise(text: string): string {
  const out: string[] = [];
  for (const ch of text) {
    const idx = NORM.get(ch);
    out.push(idx !== undefined ? LAT[idx] : ch);
  }
  return out.join("");
}

// ---- BYTE UTILITIES ----------------------------

function concatBytes(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

function xorBytes(data: Uint8Array, keystream: Uint8Array): Uint8Array {
  const out = new Uint8Array(data.length);
  for (let i = 0; i < data.length; i++) out[i] = data[i] ^ keystream[i];
  return out;
}

// ---- KEY DERIVATION (crypto.subtle) ------------

// crypto.subtle's TS types require ArrayBufferView<ArrayBuffer> in very
// recent lib.dom versions, which plain Uint8Array<ArrayBufferLike> doesn't
// satisfy. This cast is a typing-only nit (SharedArrayBuffer exclusion) --
// no behavioral effect, since these buffers are never shared.
function bs(u: Uint8Array): BufferSource {
  return u as unknown as BufferSource;
}

async function sha256(data: Uint8Array): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest("SHA-256", bs(data)));
}

async function pbkdf2(
  password: Uint8Array,
  salt: Uint8Array,
  dklenBytes: number,
  iterations: number
): Promise<Uint8Array> {
  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    bs(password),
    "PBKDF2",
    false,
    ["deriveBits"]
  );
  const bits = await crypto.subtle.deriveBits(
    { name: "PBKDF2", salt: bs(salt), iterations, hash: "SHA-256" },
    keyMaterial,
    dklenBytes * 8
  );
  return new Uint8Array(bits);
}

async function hmacSha256(keyBytes: Uint8Array, data: Uint8Array): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey(
    "raw",
    bs(keyBytes),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  return new Uint8Array(await crypto.subtle.sign("HMAC", key, bs(data)));
}

/** One PBKDF2 call -- returns 64 bytes of base key material. */
async function deriveBootstrap(password: string, cover: string): Promise<Uint8Array> {
  const enc = new TextEncoder();
  const saltSource = enc.encode(normalise(cover).slice(0, SALT_COVER_CHARS));
  const salt = await sha256(saltSource);
  const passBytes = enc.encode(password);
  const baseKey = await pbkdf2(passBytes, salt, 64, PBKDF2_ITERATIONS);
  passBytes.fill(0);
  return baseKey;
}

function bootstrapSeed(baseKey: Uint8Array): Uint8Array {
  return baseKey.slice(0, SEED_BYTES);
}

/**
 * Expand base key with nonce via HKDF-SHA256 (HMAC extract + expand).
 * Returns { seed[8], keystream[messageLen] }.
 *
 * Note: the expand-step counter is packed as a single byte, matching
 * kometa.py's `i.to_bytes(1, "big")`. Python throws OverflowError past
 * i=255; this port wraps instead (Uint8Array truncates to the low byte).
 * That's a real behavioral difference for very large messages -- flagging
 * it rather than silently hoping nobody hits it.
 */
async function deriveKeys(
  baseKey: Uint8Array,
  messageLen: number,
  nonce: Uint8Array
): Promise<{ seed: Uint8Array; keystream: Uint8Array }> {
  const keylen = Math.ceil((SEED_BYTES + messageLen) / BLOCK_SIZE) * BLOCK_SIZE;
  const prk = await hmacSha256(nonce, baseKey);

  let out: Uint8Array = new Uint8Array(0);
  let prev: Uint8Array = new Uint8Array(0);
  const iterations = Math.floor(keylen / 32) + 1;
  for (let i = 1; i <= iterations; i++) {
    const input = concatBytes(prev, new Uint8Array([i]));
    prev = await hmacSha256(prk, input);
    out = concatBytes(out, prev);
  }
  prk.fill(0);

  const key = out.slice(0, keylen);
  return { seed: key.slice(0, SEED_BYTES), keystream: key.slice(SEED_BYTES) };
}

// ---- BIT PACKING -------------------------------

function toBits(data: Uint8Array): number[] {
  if (data.length > 65535) throw new Error("Message too long (max 65,535 bytes)");
  const framed = concatBytes(
    new Uint8Array([data.length >> 8, data.length & 0xff]),
    data
  );
  const bits: number[] = [];
  for (const b of framed) {
    for (let i = 7; i >= 0; i--) bits.push((b >> i) & 1);
  }
  return bits;
}

function fromBits(bits: number[]): Uint8Array {
  const readByte = (offset: number): number => {
    let v = 0;
    for (let b = 0; b < 8; b++) {
      v = (v << 1) | (bits[offset + b] ?? 0);
    }
    return v;
  };
  const length = (readByte(0) << 8) | readByte(8);
  const out = new Uint8Array(length);
  for (let i = 0; i < length; i++) out[i] = readByte(16 + i * 8);
  return out;
}

// ---- xoshiro128** -------------------------------

function rotl(x: number, k: number): number {
  return ((x << k) | (x >>> (32 - k))) >>> 0;
}

function makeRng(seed: Uint8Array): () => number {
  const view = new DataView(seed.buffer as ArrayBuffer, seed.byteOffset, seed.byteLength);
  let s0 = (view.getUint32(0, true) | 1) >>> 0;
  let s1 = (view.getUint32(4, true) | 1) >>> 0;
  let s2 = 0x9e3779b9;
  let s3 = 0x6c62272e;

  return function rng(): number {
    // Operands here stay well under 2^53, so plain `*` keeps exact
    // integer precision -- no Math.imul needed at these magnitudes.
    const r = (rotl((s1 * 5) >>> 0, 7) * 9) >>> 0;
    const t = (s1 << 9) >>> 0;
    s2 ^= s0; s3 ^= s1; s1 ^= s2; s0 ^= s3; s2 ^= t;
    s2 >>>= 0; s3 = rotl(s3, 11);
    s0 >>>= 0; s1 >>>= 0;
    return r;
  };
}

function shuffle<T>(arr: T[], rng: () => number): T[] {
  const a = arr.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = rng() % (i + 1);
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

// ---- CARRIER POSITION SELECTION ----------------

function deadZoneBounds(n: number): { lo: number; hi: number } {
  const lo = Math.floor(n * DEAD_ZONE_START);
  const hi = n - Math.floor(n * DEAD_ZONE_END);
  return { lo, hi };
}

/** Select 64 fixed carrier positions for the nonce (drawn before header/body). */
function noncePositions(positions: number[], seed: Uint8Array): number[] {
  const { lo, hi } = deadZoneBounds(positions.length);
  const eligible = positions.slice(lo, hi);
  if (eligible.length < 64) {
    throw new Error(`Cover too small: need at least 64 carriers for nonce, got ${eligible.length}`);
  }
  return shuffle(eligible, makeRng(seed)).slice(0, 64);
}

/**
 * Select 16 fixed carrier positions for the length header. Drawn from the
 * dead-zone-excluded pool independent of message length, so decode can
 * always recover them without knowing n_bits first.
 */
function headerPositions(positions: number[], seed: Uint8Array, bs: Uint8Array): number[] {
  const { lo, hi } = deadZoneBounds(positions.length);
  const nonceSet = new Set(noncePositions(positions, bs));
  const eligible = positions.slice(lo, hi).filter((p) => !nonceSet.has(p));
  if (eligible.length < 16) {
    throw new Error(`Cover too small: need at least 16 eligible carriers, got ${eligible.length}`);
  }
  return shuffle(eligible, makeRng(seed)).slice(0, 16);
}

/** Select n_bits carrier positions for the message body via density-matched buckets. */
function bodyPositions(
  positions: number[],
  headerPos: number[],
  nBits: number,
  seed: Uint8Array,
  bs: Uint8Array,
  nBuckets: number = DENSITY_BUCKETS
): number[] {
  const { lo, hi } = deadZoneBounds(positions.length);
  const nonceSet = new Set(noncePositions(positions, bs));
  const headerSet = new Set(headerPos);
  const eligible = positions.slice(lo, hi).filter((p) => !nonceSet.has(p) && !headerSet.has(p));

  if (eligible.length < nBits) {
    throw new Error(
      `Cover too small after dead-zone trim: need ${nBits} body carriers, have ${eligible.length} eligible`
    );
  }

  const actualBuckets = Math.min(nBuckets, eligible.length);
  const bucketSize = eligible.length / actualBuckets;
  const buckets: number[][] = [];
  for (let i = 0; i < actualBuckets; i++) {
    buckets.push(eligible.slice(Math.floor(i * bucketSize), Math.floor((i + 1) * bucketSize)));
  }

  const raw = buckets.map((b) => (b.length / eligible.length) * nBits);
  const alloc = raw.map((r) => Math.floor(r));
  const remainder = nBits - alloc.reduce((a, b) => a + b, 0);

  const fracOrder = raw
    .map((r, i) => ({ i, frac: r - alloc[i] }))
    .sort((a, b) => b.frac - a.frac)
    .map((x) => x.i);
  for (const i of fracOrder.slice(0, remainder)) alloc[i] += 1;

  const rng = makeRng(seed);
  const selected: number[] = [];
  for (let i = 0; i < buckets.length; i++) {
    selected.push(...shuffle(buckets[i], rng).slice(0, alloc[i]));
  }
  return selected;
}

// ---- STEGANOGRAPHY EMBED / EXTRACT -------------

function embed(
  cover: string,
  bits: number[],
  bs: Uint8Array,
  seed: Uint8Array,
  nonce: Uint8Array
): string {
  const headerBits = bits.slice(0, 16);
  const bodyBits = bits.slice(16);

  const chars = Array.from(cover);
  const positions: number[] = [];
  chars.forEach((ch, i) => { if (LOOKUP.has(ch)) positions.push(i); });

  const noncePos = noncePositions(positions, bs);
  const hdrPos = headerPositions(positions, seed, bs);
  const bodyPos = bodyPositions(positions, hdrPos, bodyBits.length, seed, bs);

  const nonceBits: number[] = [];
  for (let i = 0; i < 64; i++) {
    nonceBits.push((nonce[Math.floor(i / 8)] >> (7 - (i % 8))) & 1);
  }

  const active = new Map<number, number>();
  noncePos.forEach((p, i) => active.set(p, nonceBits[i]));
  hdrPos.forEach((p, i) => active.set(p, headerBits[i]));
  bodyPos.forEach((p, i) => active.set(p, bodyBits[i]));

  const out = chars.map((ch, i) => {
    const entry = LOOKUP.get(ch);
    if (!entry) return ch;
    const bit = active.get(i) ?? 0; // inactive positions -> bit 0
    return bit ? CYR[entry.idx] : LAT[entry.idx];
  });
  return out.join("");
}

function extractNonce(encoded: string, bs: Uint8Array): Uint8Array {
  const chars = Array.from(encoded);
  const positions: number[] = [];
  const observed = new Set<number>();
  chars.forEach((ch, i) => {
    const entry = LOOKUP.get(ch);
    if (entry) {
      positions.push(i);
      if (entry.script === "cyr") observed.add(i);
    }
  });

  const noncePos = noncePositions(positions, bs);
  const bits = noncePos.map((p) => (observed.has(p) ? 1 : 0));
  const bytes = new Uint8Array(8);
  for (let i = 0; i < 8; i++) {
    let v = 0;
    for (let b = 0; b < 8; b++) v = (v << 1) | bits[i * 8 + b];
    bytes[i] = v;
  }
  return bytes;
}

function extract(encoded: string, seed: Uint8Array, bs: Uint8Array): number[] {
  const chars = Array.from(encoded);
  const positions: number[] = [];
  const observed = new Map<number, number>();
  chars.forEach((ch, i) => {
    const entry = LOOKUP.get(ch);
    if (entry) {
      positions.push(i);
      if (entry.script === "cyr") observed.set(i, 1);
    }
  });

  const readBit = (p: number) => observed.get(p) ?? 0;
  const readByte = (ps: number[], offset: number) => {
    let v = 0;
    for (let b = 0; b < 8; b++) v = (v << 1) | readBit(ps[offset + b]);
    return v;
  };

  const hdrPos = headerPositions(positions, seed, bs);
  const payloadBytes = (readByte(hdrPos, 0) << 8) | readByte(hdrPos, 8);
  let nBodyBits = payloadBytes * 8;

  const { lo, hi } = deadZoneBounds(positions.length);
  const nonceSet = new Set(noncePositions(positions, bs));
  const hdrSet = new Set(hdrPos);
  const maxBodyBits = positions.slice(lo, hi).filter((p) => !nonceSet.has(p) && !hdrSet.has(p)).length;
  nBodyBits = Math.min(nBodyBits, Math.max(0, maxBodyBits));

  const bodyPos = bodyPositions(positions, hdrPos, nBodyBits, seed, bs);
  const headerBits = hdrPos.map(readBit);
  const bodyBits = bodyPos.map(readBit);
  return headerBits.concat(bodyBits);
}

// ---- PUBLIC API ---------------------------------

/** Encode `message` into `cover` text using `password`. Returns the carrier text. */
export async function encode(cover: string, message: Uint8Array, password: string): Promise<string> {
  const trimmedCover = cover.replace(/\s+$/, "");
  const nonce = crypto.getRandomValues(new Uint8Array(8));

  const baseKey = await deriveBootstrap(password, trimmedCover);
  const bs = bootstrapSeed(baseKey);
  const { seed, keystream } = await deriveKeys(baseKey, message.length, nonce);

  const encrypted = xorBytes(message, keystream);
  const bits = toBits(encrypted);
  const output = embed(trimmedCover, bits, bs, seed, nonce);

  baseKey.fill(0);
  keystream.fill(0);
  seed.fill(0);

  return output;
}

/** Decode a carrier string produced by `encode`, given the same `password`. */
export async function decodeText(encoded: string, password: string): Promise<Uint8Array> {
  const trimmed = encoded.replace(/\s+$/, "");

  const baseKey = await deriveBootstrap(password, trimmed);
  const bs = bootstrapSeed(baseKey);
  const nonce = extractNonce(trimmed, bs);

  const first = await deriveKeys(baseKey, BLOCK_SIZE, nonce); // seed only, dummy length
  const bits = extract(trimmed, first.seed, bs);
  const encrypted = fromBits(bits);

  const second = await deriveKeys(baseKey, encrypted.length, nonce); // real keystream
  const result = xorBytes(encrypted, second.keystream);

  baseKey.fill(0);
  first.seed.fill(0);
  first.keystream.fill(0);
  second.seed.fill(0);
  second.keystream.fill(0);

  return result;
}

// Convenience wrappers for string messages (UTF-8 in/out).
export async function encodeText(cover: string, message: string, password: string): Promise<string> {
  return encode(cover, new TextEncoder().encode(message), password);
}

export async function decodeToString(encoded: string, password: string): Promise<string> {
  return new TextDecoder().decode(await decodeText(encoded, password));
}